### ISSUE TITLE
Change VF NIC definition from hostdev to interface of type hostdev

### ISSUE THREAD

**Issue by atodorov-storpool** [OpenNebula/one#3329]
**Description**

The VF NIC interfaces are created with the common VF hostdev definition, like:

```xml
        <hostdev mode='subsystem' type='pci' managed='yes'>
            <source>
                <address  domain='0x0000' bus='0xd8' slot='0x00' function='0x4'/>
            </source>
            <address type='pci' domain='0x0000' bus='0x01' slot='0x01' function='0'/>
        </hostdev>
```

This configuration does not define the VF NIC's MAC address leaving this task to the Guest's kernel. So on each VM start the VM has different(generated) MAC address.

Switching to _interface_ definition of type _hostdev_ allow OpenNebula to provide a persistent MAC address(generated by OpenNebula or provided in the VM's template?) to the VF NIC associated with the given VM, like:

```xml
        <interface managed="yes" type="hostdev">
            <driver name="vfio" />
            <mac address="6e:86:82:68:b8:47" />
            <source>
                <address bus="0xd8" domain="0x0000" function="0x4" slot="0x00" type="pci" />
            </source>
            <address bus="0x01" domain="0x0000" function="0" slot="0x01" type="pci" />
        </interface>
```

**Use case**

* The VM's could have proprietary software that is tied to MAC address
* The vm could be configured to PXE boot from the VF NIC and the associated bootp/dhcp server could require persistent VF NIC

**Interface Changes**

Change the core to generate VF NIC configuration that allow setting a MAC address

**Additional Context**

IMHO OpenNebula should generate a VF NIC mac if it is not provided in the VM's Template... (for each VF NIC attached)

## Progress Status
- [ ] Branch created 
- [ ] Code committed to development branch
- [ ] Testing - QA
- [ ] Documentation 
- [ ] Release notes - resolved issues, compatibility, known issues
- [ ] Code committed to upstream release/hotfix branches
- [ ] Documentation committed to upstream release/hotfix branches

**Comment by rsmontero** [comment]
This is handled by the context packages, [network-pci script](https://github.com/OpenNebula/addon-context-linux/blob/master/src/etc/one-context.d/loc-10-network-pci%23%23one). Guest is able to address and identify the NIC by its PCI address.

So OpenNebula actually sets the VM's MAC address as the one in the template. So I'd say that the first use-case is fully implemented.

During boot the VM will have a transient MAC address, though. Note that the method that currently uses OpenNebula works for both PF and VF; and your proposed solution will not work with single root devices. 

Also in 5.10 we have the network driver hooks that could set the MAC address in that cases you need the specific MAC before the VM executes the contextualization. (Doing something like: #ip link set <pf> vf <vf_index> mac <vnic_mac> before starting the VM)

**Comment by atodorov-storpool** [comment]
> So OpenNebula actually sets the VM's MAC address as the one in the template. So I'd say that the first use-case is fully implemented.

OpenNebula allowed to define a NIC PCI device that is not associated to a VNet:
```xml
  <TEMPLATE>
    <NIC>
      <NETWORK><![CDATA[net-10.2.1]]></NETWORK>
      <NETWORK_UNAME><![CDATA[ant]]></NETWORK_UNAME>
      <SECURITY_GROUPS><![CDATA[0]]></SECURITY_GROUPS>
    </NIC>
    <NIC_DEFAULT>
      <MODEL><![CDATA[virtio]]></MODEL>
    </NIC_DEFAULT>
    <PCI>
      <CLASS><![CDATA[0200]]></CLASS>
      <DEVICE><![CDATA[1018]]></DEVICE>
      <TYPE><![CDATA[NIC]]></TYPE>
      <VENDOR><![CDATA[15b3]]></VENDOR>
    </PCI>
  </TEMPLATE>
```
I'll run some tests next week with VNet assigned to see what is the delay and is there impact on *deb OS-es :)


> During boot the VM will have a transient MAC address, though. Note that the method that currently uses OpenNebula works for both PF and VF; and your proposed solution will not work with single root devices.

What do you think of an option to set the type of the PCI device PF/VF in the filter or in core? Then create a different definition for VF's?

My concern is how it will affect Windows guests or other OS guests with DRM/licensing linked to the NIC's mac. With persistent MAC there are no issue with that.

> Also in 5.10 we have the network driver hooks that could set the MAC address in that cases you need the specific MAC before the VM executes the contextualization. (Doing something like: #ip link set vf <vf_index> mac <vnic_mac> before starting the VM)

Alex shared info regarding the new development so I am eager to test it and see how many patches will drop from the addon-storpool ;) In the alternate [deployment script](https://github.com/OpenNebula/addon-storpool/blob/master/vmm/kvm/deploy-tweaks) that I use there is already a [tweak](https://github.com/OpenNebula/addon-storpool/blob/master/vmm/kvm/deploy-tweaks.d.example/vfhostdev2interface.py) that do fix the domain XML. Will extend it to use the MAC generated by ONE (after changing the VM template, though)

Cheers

**Comment by rsmontero** [comment]
Implemented as part of #5940
